### PR TITLE
Lazy Tensor Simulation

### LINKED ISSUE #138: [Feature Request]  Speed up simulation by removing patches that are not on the lattice from the state vector
Issue Description
-----------------
Simulation is a critical bottleneck right now and it is because of simulating useless patches.

Right now simulation of longer circuits happens at a much slower speed because every patch that is ever on the lattice contributes to one big state vector. In particular state vector size grows exponentially with y-depth and t-depth, because each of these operations is associated with an new patch that contributes an extra qubit in the logical lattice computation, doubling it's size.

Even these patches aren't contributing to the computation for nearly the whole length of the computation. This makes simulating circuits of just a couple lines infeasible without y operator removal.

Addressing this issue will greatly expand the set of circuits we can test.

Proposed Solution
-----------------

Here are some possible solutions:
- Exclude patches not on the lattice from the state vector
  - Might be the easiest to implement 
- Break the global state into smaller vectors based on separability
   - We already have a separability algorithm
   - Can cache what was not entangled from previous operations  
   - Speeds up all not entangled computations (E.g. different modules running in parallel)
- Leverage an existing simulator to simulate one operation at a time

### PR DESCRIPTION
**Pull request by gwwatkin**
If we have a state like:

`|0> otime |1> otimes (|00>+|11>)/sqrt(2) otimes |0>`

We don't need to expanded out to a state vector with 32 entries to apply, say an X gate to the second qubit.

This PR introduces simulation that reduces the size of the state vectors involved in simulation by tracking the list of tensors operands that make up the tensor state. The class used to hold these tensored operands states is `LazyTensorOps` introduced in #213.

Operations supported:
 - [x] Single qubit gates
 - [x] Single qubit measurements
 - [x] Multibody measurements

Closes #138

### CODE REVIEW

**Comment by gwwatkin** [comment]
It seems like there is some subtlety about qubit indices appearing as I write tests ... unfortunately will have to address that in this PR because of the large amount of tests of changes :sweat:

**Comment by gwwatkin** [comment]
The tests aren't as fine grained as uni tests should be, as each method's test require other classes to test, but I'm fine with this situation as I think it still gives us good compromise between time spent writing tests and precision of the tests

**Comment by gwwatkin** [comment]
@isolatedinformation  and @alexnguyenn Ready for a first round of review and comments ... unfortunately the bookkeeping for multi-body measurements got out of hand so this PR is far from clean. I'm happy to go over it in a meeting as well

File: src/lsqecc/pauli_rotations/circuit.py [hunk 1]
```diff
@@ -321,6 +321,14 @@ def load_from_pyzx(circuit) -> "PauliOpCircuit":
         print("Gate Missed: ", gate_missed)
         return ret_circ
 
+    @staticmethod
+    def from_list(pauli_op_list: List[PauliProductOperation]):
+        c = PauliOpCircuit(pauli_op_list[0].qubit_num)
+        for op in pauli_op_list:
+            assert c.qubit_num == op.qubit_num
```
Comment (isolatedinformation) [inline src/lsqecc/pauli_rotations/circuit.py:328]: Minor Suggestion: I think we should call it `num_qubits` instead `qubit_num`, because later might thought of as returning the index. There is scope for ambiguity. I see it's called `num_qubits` in line 177 in `lazy_tensor_op.py`

Comment (gwwatkin) [reply src/lsqecc/pauli_rotations/circuit.py:328]: yep we definitely should ... I think this PR is already big enough though

File: src/lsqecc/simulation/logical_patch_state_simulation.py [hunk 2]
```diff
@@ -263,3 +292,184 @@ def apply_logical_operation(self, logical_op: llops.LogicalLatticeOperation):
...
+        ]
+
+        involved_operand_idxs: List[int] = list(
+            set(
+                [
+                    self.logical_state.get_idxs_of_qubit(qubit_idx)[0]
+                    for qubit_idx in involved_qubit_idxs
+                ]
+            )
+        )
```
Comment (alexnguyenn) [inline src/lsqecc/simulation/logical_patch_state_simulation.py]: I am confused here - `list` to `set` to `list`?

File: src/lsqecc/simulation/lazy_tensor_op.py [hunk 3]
```diff
@@ -38,7 +39,7 @@ class LazyTensorOp(Generic[T]):
     operands. Has methods to apply operators so that the tensor structure is preserved."""
 
     def __init__(self, ops: List[T]):
-        self.ops = ops
+        self.ops = [op for op in ops]  # TODO rename to operands
```
Comment (alexnguyenn) [inline src/lsqecc/simulation/lazy_tensor_op.py]:
```suggestion
        self.ops = ops[:]  # TODO rename to operands
```
I think something like this also works as well.

File: src/lsqecc/simulation/logical_patch_state_simulation.py [hunk 4]
```diff
@@ -263,3 +292,184 @@ def apply_logical_operation(self, logical_op: llops.LogicalLatticeOperation):
...
+            )
+            self.logical_state.ops[operand_idx] = cast(qkop.StateFn, symbolic_state.eval())
+
+        elif isinstance(logical_op, llops.MultiBodyMeasurement):
+            # Prepare the state by moving all involved operands to the front
+            self.align_state_with_pauli_op_to_first_operand(logical_op)
+            size_of_first_operand = self.logical_state.ops[0].num_qubits
+
+            # Make a global observable for the operator
+            global_observable = qkop.I ^ size_of_first_operand
```
Comment (isolatedinformation) [inline src/lsqecc/simulation/logical_patch_state_simulation.py]: After discussion with @gwwatkin, for Lazy Tensor this means `first_operand_observable`, with the first_operand effectively causing the multibody measurement . `global_observable` might be misleading

File: src/lsqecc/simulation/lazy_tensor_op.py [hunk 5]
```diff
@@ -77,23 +78,111 @@ def apply_more_granular_lazy_tensor(
...
+            f"Requesting qubit index: {qubit_idx} out of range: {operand_boundaries[-1]}"
+        )
+
+    def get_idx_of_first_qubit_in_operand(self, operand_idx: int):
+        return self.get_idx_of_first_qubit_for_each_operand()[operand_idx]
+
+    def swap_operands(self, operand_idx1: int, operand_idx2: int) -> None:
+        tmp = self.ops[operand_idx2]
+        self.ops[operand_idx2] = self.ops[operand_idx1]
+        self.ops[operand_idx1] = tmp
```
Comment (alexnguyenn) [inline src/lsqecc/simulation/lazy_tensor_op.py]:
```suggestion
        self.ops[operand_idx2], self.ops[operand_idx1] = self.ops[operand_idx1], self.ops[operand_idx2]
```
You can do something like this in Python

File: src/lsqecc/simulation/logical_patch_state_simulation.py [hunk 6]
```diff
@@ -263,3 +292,184 @@ def apply_logical_operation(self, logical_op: llops.LogicalLatticeOperation):
...
+            get_init_state(quuid) for idx, quuid in self.mapper.enumerate_patches_by_index()
+        ]
+        return LazyTensorOp(all_init_states)
+
+    def apply_logical_operation(self, logical_op: llops.LogicalLatticeOperation):
+        """Update the logical state"""
+
+        if not logical_op.does_evaluate():
+            raise Exception(
+                "apply_logical_operation called with non evaluating operation :" + repr(logical_op)
```
Comment (alexnguyenn) [inline src/lsqecc/simulation/logical_patch_state_simulation.py]:
```suggestion
                f"apply_logical_operation called with non evaluating operation: {repr(logical_op)}"
```

**Review by alexnguyenn — COMMENTED** [review]
I put some comments but I think we will need a meeting to go over all of the logic change.

File: src/lsqecc/simulation/logical_patch_state_simulation.py (hunk 2 above)
Comment (gwwatkin) [reply src/lsqecc/simulation/logical_patch_state_simulation.py]: yes to remove duplicates

Comment (gwwatkin) [reply src/lsqecc/simulation/logical_patch_state_simulation.py]: Though it looks like it's no longer necessary

**Comment by alexnguyenn** [comment]
pre-commit.ci autofix

**Comment by gwwatkin** [comment]
> I put some comments but I think we will need a meeting to go over all of the logic change.


@alexnguyenn Yes we probably do. Went over the changes with @isolatedinformation, but we can go over them again

File: src/lsqecc/simulation/logical_patch_state_simulation.py [hunk 7]
```diff
@@ -184,14 +186,129 @@ def _get_all_operating_patches(
...
+    def make_simulator(
+        simulator_type: SimulatorType, logical_computation: llops.LogicalLatticeComputation
+    ):
+        if simulator_type == SimulatorType.LAZY_TENSOR:
+            return LazyTensorPatchSimulator(logical_computation)
+        else:
+            return FullStateVectorPatchSimulator(logical_computation)
+
+
+class FullStateVectorPatchSimulator(PatchSimulator):
```
Comment (gwwatkin) [inline src/lsqecc/simulation/logical_patch_state_simulation.py:213]: This is just adding back the existing state vector simulator, unfortnuately it didn't occur to me that we would probably want to keep both in the end

File: src/lsqecc/simulation/logical_patch_state_simulation.py [hunk 8]
```diff
@@ -307,16 +424,52 @@ def bring_active_operands_to_front(self, logical_op: llops.MultiBodyMeasurement)
...
+                operand_uuid_list[operand_swap_target_counter],
+            )
+            self.mapper.patch_location_to_logical_idx = dict(
+                [
+                    (patch_uuid, idx)
+                    for idx, patch_uuid in enumerate(itertools.chain(*operand_uuid_list))
+                ]
+            )
+
+            operand_swap_target_counter += 1
```
Comment (gwwatkin) [inline src/lsqecc/simulation/logical_patch_state_simulation.py:449]: Looks like I didn't make a separate commit for this change or lost it in rebase, but all this is doing is fixing a mismatch of indices when swapping operators

File: src/lsqecc/simulation/logical_patch_state_simulation.py (hunk 4 above)
Comment (gwwatkin) [reply src/lsqecc/simulation/logical_patch_state_simulation.py]: Applied suggestion

**Review by isolatedinformation — APPROVED** [review]
